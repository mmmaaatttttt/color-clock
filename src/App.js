import React from 'react';
import ColorClock from "./ColorClock";

function App() {
  return (
    <ColorClock />
  );
}

export default App;
