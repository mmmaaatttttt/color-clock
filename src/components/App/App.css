.App {
  position: relative;
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column-reverse;
}

@keyframes slidein_about {
  0% {
    opacity: 0;
    transform: translateY(0%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(100%);
  }
}

@keyframes slidein_form_from_clock {
  0% {
    opacity: 0;
    transform: translateY(-100%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(0%);
  }
}

@keyframes slidein_form_from_about {
  0% {
    opacity: 0;
    transform: translateY(100%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(0%);
  }
}

@keyframes slideout_form_to_about {
  0% {
    opacity: 1;
    transform: translateY(0%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(100%);
  }
}

@keyframes slideout_about {
  0% {
    opacity: 1;
    transform: translateY(100%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(0%);
  }
}


@keyframes slideout_form_to_clock {
  0% {
    opacity: 1;
    transform: translateY(0%);
  }

  80% {
    opacity: 0;
  }

  100% {
    transform: translateY(-100%);
  }
}


@keyframes slidein_clock {
  0% {
    opacity: 0;
    transform: translateY(200%);
  }

  80% {
    opacity: 1;
  }

  100% {
    transform: translateY(100%);
  }
}

.page-exit-active .ClockForm {
  animation: slideout_form_to_about 1000ms forwards;
}

.page-enter-active .ClockForm {
  animation: slidein_form_from_about 1000ms forwards;
}

.page-enter-active .About {
  animation: slidein_about 1000ms forwards;
}

.page-exit-active .About  {
  animation: slideout_about 1000ms forwards;
}

.page-enter-active .ClockForm.ClockForm-animate-up {
  animation: slidein_form_from_clock 1000ms forwards;
}

.page-exit-active .ClockForm.ClockForm-animate-up {
  animation: slideout_form_to_clock 1000ms forwards;
}

.page-enter-active .ColorClock { 
  animation: slidein_clock 1000ms forwards;
}
