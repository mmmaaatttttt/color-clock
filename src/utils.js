/**
 * Capitalizes the first letter in a string.
 *
 * @param {String} str - String to be capitalized.
 * @returns {String} Returns the capitalized string.
 */
export function capitalize(str) {
  return str[0].toUpperCase() + str.slice(1);
}

/**
 * Converts a time of the form "HH:MM" or "HH:MM:SS"
 * to a number in seconds.
 *
 * @param {String} time - Time in one of the above formats.
 * @returns {Number} Returns a number in seconds.
 */
export function timeToNumber(time) {
  const secondsPerHour = 3600;
  const secondsPerMinute = secondsPerHour / 60;
  let timeArr = time.split(":").map(Number);

  // if no seconds provided, default to 0 seconds
  if (timeArr.length === 2) timeArr.push(0);

  return (
    timeArr[0] * secondsPerHour + timeArr[1] * secondsPerMinute + timeArr[2]
  );
}
