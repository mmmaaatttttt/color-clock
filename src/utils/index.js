import moment from "moment";

/**
 * Capitalizes the first letter in a string.
 *
 * @param {String} str - String to be capitalized.
 * @returns {String} Returns the capitalized string.
 */
export function capitalize(str) {
  if (!str) return "";
  return str[0].toUpperCase() + str.slice(1);
}

/**
 * Converts a time of the form "HH:MM" or "HH:MM:SS"
 * to a number in seconds.
 *
 * @param {String} time - Time in one of the above formats.
 * @returns {Number} Returns a number in seconds.
 */
export function timeToNumber(time) {
  const secondsPerHour = 3600;
  const secondsPerMinute = secondsPerHour / 60;
  let timeArr = time.split(":").map(Number);

  // if no seconds provided, default to 0 seconds
  if (timeArr.length === 2) timeArr.push(0);

  return (
    timeArr[0] * secondsPerHour + timeArr[1] * secondsPerMinute + timeArr[2]
  );
}

/**
 * Gets the current time; returns an object with the time in two formats.
 *
 * @returns {Object} an object with two formats for the time:
 *  timeForDisplay: h:mm A (e.g. 3:16 PM)
 *  timeForColor: H:mm:ss (e.g. 15:16:23)
 *
 * The first formatting is for the view,
 * the second is for determining the background color and general comparison
 * (since start and end times are in 24 hour formatting).
 */
export function getTime() {
  const newMoment = moment();

  return {
    currentTimeForDisplay: newMoment.format("LT"),
    currentTimeForComparison: newMoment.format("HH:mm:ss")
  };
}

/**
 * Generate a human readable description of how long ago the input time was
 * (for times within the same day).
 *
 * @param {String} time - 24 hour time as a string, e.g. "15:16"
 * @returns {String} English description of how long ago the time was.
 */
export function howLongFromNow(time) {
  const [hour, minutes] = time.split(":").map(Number);
  return moment()
    .startOf("day")
    .add(hour, "hours")
    .add(minutes, "minutes")
    .fromNow();
}

/**
 * Calculate time2 as a percentage of the duration between time1 and time3.
 *
 * @param {String} time1 - String representation of start time, e.g. "08:00"
 * @param {String} time2 - String representation of start time, e.g. "08:30"
 * @param {String} time3 - String representation of start time, e.g. "08:45"
 * @returns {Number} time2's percentage
 */
export function middleTimeAsPercentage(time1, time2, time3) {
  if (time1 >= time2 || time2 >= time3) {
    throw new Error(
      `Times must be provided in increasing order! Received ${time1}, ${time2}, ${time3}.`
    );
  }

  const [time1Num, time2Num, time3Num] = [time1, time2, time3].map(
    timeToNumber
  );

  return ((time2Num - time1Num) * 100) / (time3Num - time1Num);
}
