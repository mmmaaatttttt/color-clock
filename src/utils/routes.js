import PowerRadix from "power-radix";
import { timeToNumber } from "./";

const charSet = "23456789abcdefghijkmnpqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ".split(
  ""
);
const hexSet = "0123456789abcdef".split("");
const hexRadix = hexSet.length;

/**
 * Convert and object containing time data (24-hour format)
 * and color data (hex codes) into a base-56 string.
 * Uses the characters in `charSet` for the encoding for URL-friendliness.
 * 
 * This function also validates that the colors are valid hex codes,
 * the times are in a valid format, the times are ascending in the array,
 * and that the number of times passed in matches the number of colors.
 * 
 * @param {Object} colorsAndTimesObj - object of color and time data
 * @param {String[]} colorsAndTimesArr.times - Array of time strings
 * @param {String[]} colorsAndTimesArr.colors - Array of color hex strings
 * @returns {String} Encoded string
 */
export function encodeToUrl(colorsAndTimesObj) {
  let colorsAndTimesArr = [];
  if (_validate(colorsAndTimesObj)) {
    for (let time of colorsAndTimesObj.times) {
      colorsAndTimesArr.push(_timeToHex(time));
    }

    for (let color of colorsAndTimesObj.colors) {
      colorsAndTimesArr.push(color.replace("#", "").toLowerCase());
    }

    return new PowerRadix(colorsAndTimesArr.join(""), hexSet).toString(charSet);
  }
}

/**
 * Inverse of encodeToUrl - takes an encoded string,
 * returns an object of color and time data.
 * 
 * As with encodeToUrl, this function will throw errors
 * if the decoded data is invalid, or if the decoding process fails.
 * 
 * @param {String} path - encoded string extracted from a URL.
 * @returns {Object} Object of colors / times data.
 */
export function decodeFromUrl(path) {
  let urlToHexStr = new PowerRadix(path, charSet).toString(hexSet);
  let charsPerTime = 3;
  let charsPerColor = 6;
  let charsPerPair = charsPerTime + charsPerColor;
  let numTimesAndColors = Math.ceil(urlToHexStr.length / charsPerPair);

  // handle case of leading zeroes
  let paddedUrlToHexStr = urlToHexStr.padStart(
    numTimesAndColors * charsPerPair,
    "0"
  );

  let times = [];
  let colors = [];

  for (let i = 0; i < numTimesAndColors; i++) {
    let timeIdx = i * charsPerTime;
    let timeHex = paddedUrlToHexStr.slice(timeIdx, timeIdx + charsPerTime);
    times.push(_hexToTime(timeHex));

    let colorIdx = numTimesAndColors * charsPerTime + i * charsPerColor;
    let colorHex = paddedUrlToHexStr.slice(colorIdx, colorIdx + charsPerColor);
    colors.push(`#${colorHex}`);
  }

  let colorsAndTimesObj = { colors, times };
  if (_validate(colorsAndTimesObj)) {
    return colorsAndTimesObj;
  }
}

/**
 * Validation helper for decodeFromUrl / encodeToUrl.
 * Ensures that all color / time data to be encoded is valid.
 * 
 * @param {Object} param0 - Object with array of times and array of colors 
 * @param {String[]} param0.times - Array of time strings
 * @param {String[]} param0.colors - Array of color hex strings
 * @returns {Boolean} true if no errors, throws if errors occur
 */
function _validate({ times, colors }) {
  if (times.length !== colors.length) {
    throw new Error(
      `Must provide same number of colors and times. ${colors.length} (colors length) !== ${times.length} (times length).`
    );
  }

  const allTimesValid = times.every(_validTime);
  const allColorsValid = colors.every(_validColor);
  const timesInOrder = times.slice(1).every((time, idx) => {
    const lastTime = times[idx];
    return lastTime < time;
  });

  if (!allTimesValid) {
    throw new Error(`Times must be in valid HH:mm format. Received ${times.join(", ")}.`);
  }

  if (!allColorsValid) {
    throw new Error(
      `Colors must be valid hex codes with a leading '#'. Received ${colors.join(", ")}.`
    );
  }

  if (!timesInOrder) {
    throw new Error(`Times must be in ascending order. Received ${times.join(", ")}.`);
  }

  return true;
}

/**
 * Regex to check whether a time matches the HH:mm pattern
 * (HH must be between 00 and 23, mm must be between 0 and 59).
 * 
 * @param {String} timeStr - time string to validate
 * @returns {Boolean} true if timeStr is a valid time, false otherwise
 */
function _validTime(timeStr) {
  return /^([01]?[0-9]|2[0-3]):[0-5][0-9]$/.test(timeStr);
}

/**
 * Regex to check whether a color is a valid hex code.
 * 
 * @param {String} colorStr - color string to validate
 * @returns {Boolean} true if colorStr is a valid color, false otherwise
 */
function _validColor(colorStr) {
  return /^#[0-9A-F]{6}$/i.test(colorStr);
}

/**
 * Our encoding works by first converting the times into hex codes,
 * then combining these with the color hexes to create a long base-16 string.
 * The encoding then just moves from base-16 to base-56.
 * 
 * This is a helper to conver a time string into a hexadecimal string.
 * 
 * @param {String} time - time string
 * @returns {String} hexadecimal representation of the time
 */
function _timeToHex(time) {
  const secondsPerMinute = 60;
  return (timeToNumber(time) / secondsPerMinute)
    .toString(hexRadix)
    .toLowerCase()
    .padStart(3, "0");
}

/**
 * This function is the inverse of _timeToHex, described above.
 * 
 * @param {String} hex - hexadecimal string
 * @returns {String} time string
 */
function _hexToTime(hex) {
  const minutesPerHour = 60;
  let hexAsNum = parseInt(hex, hexRadix);
  const minutes = (hexAsNum % minutesPerHour).toString().padStart(2, "0");
  const hour = Math.floor(hexAsNum / minutesPerHour)
    .toString()
    .padStart(2, "0");
  return `${hour}:${minutes}`;
}
