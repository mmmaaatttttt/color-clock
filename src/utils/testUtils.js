import React from "react";
import MockDate from "mockdate";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render } from "@testing-library/react";

/**
 * Create a timestamp out of a 24 hour time string.
 * Used in conjunction with mocking the Date constructor via MockDate.
 *
 * @param {String} time - 24 hour time in HH:mm:ss format (e.g. 15:16:23)
 */
function makeTimeStamp(time) {
  return `Sun Jul 05 2020 ${time} GMT-0700`;
}

/**
 * Abstraction around MockDate that lets us mock dates by providing
 * only times, not dates.
 */
export const timeMock = {
  set(time) {
    MockDate.set(makeTimeStamp(time));
  },
  reset() {
    MockDate.reset();
  }
};

// taken from https://testing-library.com/docs/example-react-router
export function renderWithRouter(
  component,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  const Wrapper = ({ children }) => (
    <Router history={history}>{children}</Router>
  );
  return {
    ...render(component, { wrapper: Wrapper }),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history
  };
}
