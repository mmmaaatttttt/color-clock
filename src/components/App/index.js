import React from "react";
import Routes from "../Routes";
import "./App.css";

/** 
 * Root of our react application. Related stylesheet includes
 * all keyframes for page transitions.
 * 
 * Props: none
 * State: none
 */
function App() {
  return (
    <div className="App">
      <Routes />
    </div>
  );
}

export default React.memo(App);
