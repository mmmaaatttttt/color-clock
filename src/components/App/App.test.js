import React from "react";
import App from "./";
import { renderWithRouter, timeMock } from "../../utils/testUtils";

describe("basic App tests", function () {
  it("renders without crashing", function () {
    renderWithRouter(<App />);
  });

  it("matches snapshot for the form", function () {
    const { asFragment } = renderWithRouter(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for about", function () {
    const { asFragment } = renderWithRouter(<App />, { route: "/about" });
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot for a valid clock", function () {
    timeMock.set("10:00");
    const { asFragment } = renderWithRouter(<App />, {
      route: "/clocks/3tH42XskKAkY7NYHrms"
    });
    expect(asFragment()).toMatchSnapshot();
    timeMock.reset();
  });
});
