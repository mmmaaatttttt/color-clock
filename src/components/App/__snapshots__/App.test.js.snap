// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component basic tests matches snapshot for a valid clock 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <div
        class="ColorClock ColorClock--active"
        style="background-color: rgb(128, 255, 0); box-shadow: rgb(128, 255, 0) 0 -5rem 10rem 5rem;"
      >
        <div>
          <h1>
            10:00 AM
          </h1>
          <div
            class="ColorClock--icon-wrapper"
          >
            <div
              class="Icon"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-eye-slash fa-w-20 fa-fw "
                data-icon="eye-slash"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="Icon"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-redo fa-w-16 fa-fw "
                data-icon="redo"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M500.33 0h-47.41a12 12 0 0 0-12 12.57l4 82.76A247.42 247.42 0 0 0 256 8C119.34 8 7.9 119.53 8 256.19 8.1 393.07 119.1 504 256 504a247.1 247.1 0 0 0 166.18-63.91 12 12 0 0 0 .48-17.43l-34-34a12 12 0 0 0-16.38-.55A176 176 0 1 1 402.1 157.8l-101.53-4.87a12 12 0 0 0-12.57 12v47.41a12 12 0 0 0 12 12h200.33a12 12 0 0 0 12-12V12a12 12 0 0 0-12-12z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="ProgressBar "
          data-testid="ProgressBar"
        >
          <div
            class="ProgressBar--bar"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-caret-up fa-w-10 fa-2x ProgressBar--caret"
              data-icon="caret-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="left: 33.333333333333336%;"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App component basic tests matches snapshot for about 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <div
        class="About"
      >
        <h2>
          Make time visible.
        </h2>
        <p>
          For folks who can't tell time and have a hard time with transitions (e.g. small children), color clocks give a clear visual that tells them how close they are to an upcoming transition. Time-telling skills not required.
        </p>
        <p>
          Color clocks smoothly transition between three different phases: start, warning and end. You control what time each phase should start and what each phase should look like.
        </p>
        <p>
          How to make a color clock:
        </p>
        <ol>
          <li>
            Click 
            <a
              href="/"
            >
              here
            </a>
             to fill out the color clock form.
          </li>
          <li>
            Select a time for the clock to start.
          </li>
          <li>
            Select a color for the start time (defaults to 
            <span
              style="background-color: rgb(122, 231, 185);"
            >
              green
            </span>
            ).
          </li>
          <li>
            Repeat steps 2 and 3 for the warning time and the ending time (colors default to 
            <span
              style="background-color: rgb(255, 236, 158);"
            >
              yellow
            </span>
             and 
            <span
              style="background-color: rgb(254, 173, 205);"
            >
              red
            </span>
            , respectively).
          </li>
          <li>
            Click on the button to create your clock.
          </li>
          <li>
            (Optional.) Bookmark the clock page to revisit it any day in the future. The clock will remember the times and colors you selected!
          </li>
        </ol>
        <p>
          Please note that times must all be for the same calendar day. For example, it's not possible to make a clock that starts at 11:30 PM and ends at 1:00 AM the next day.
        </p>
        <p>
          <a
            href="/"
          >
            Go back
          </a>
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App component basic tests matches snapshot for the form 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <div
        class="ClockForm"
      >
        <div>
          <svg
            height="0"
            width="0"
          >
            <lineargradient
              id="clockgrad"
              x1="0%"
              x2="100%"
              y1="0%"
              y2="100%"
            >
              <stop
                offset="0%"
                style="stop-color: rgb(122, 231, 185); stop-opacity: 1;"
              />
              <stop
                offset="50%"
                style="stop-color: rgb(255, 236, 158); stop-opacity: 1;"
              />
              <stop
                offset="100%"
                style="stop-color: rgb(254, 173, 205); stop-opacity: 1;"
              />
            </lineargradient>
          </svg>
          <h1
            class="Title--heading"
          >
            Color
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-clock fa-w-16 fa-2x Title--icon"
              data-icon="clock"
              data-prefix="far"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm61.8-104.4l-84.9-61.7c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v141.7l66.8 48.6c5.4 3.9 6.5 11.4 2.6 16.8L334.6 349c-3.9 5.3-11.4 6.5-16.8 2.6z"
                fill="currentColor"
              />
            </svg>
            Clock
          </h1>
          <a
            class="Title--question"
            href="/about"
          >
            <div
              class="Icon"
              role="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-question fa-w-12 fa-fw "
                data-icon="question"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M202.021 0C122.202 0 70.503 32.703 29.914 91.026c-7.363 10.58-5.093 25.086 5.178 32.874l43.138 32.709c10.373 7.865 25.132 6.026 33.253-4.148 25.049-31.381 43.63-49.449 82.757-49.449 30.764 0 68.816 19.799 68.816 49.631 0 22.552-18.617 34.134-48.993 51.164-35.423 19.86-82.299 44.576-82.299 106.405V320c0 13.255 10.745 24 24 24h72.471c13.255 0 24-10.745 24-24v-5.773c0-42.86 125.268-44.645 125.268-160.627C377.504 66.256 286.902 0 202.021 0zM192 373.459c-38.196 0-69.271 31.075-69.271 69.271 0 38.195 31.075 69.27 69.271 69.27s69.271-31.075 69.271-69.271-31.075-69.27-69.271-69.27z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </a>
        </div>
        <form
          class="ClockForm--form"
        >
          <div>
            <h4
              class="ClockFormRow--heading"
            >
              When do you want the clock to start?
            </h4>
            <div
              class="ClockFormRow row "
            >
              <div
                class="column time"
              >
                <label
                  for="startTime"
                >
                  Start Time
                </label>
                <input
                  id="startTime"
                  name="startTime"
                  type="time"
                  value=""
                />
              </div>
              <div
                class="column"
              >
                <label
                  for="startColor"
                >
                  Start Color
                </label>
                <input
                  id="startColor"
                  name="startColor"
                  type="color"
                  value="#7AE7B9"
                />
              </div>
            </div>
          </div>
          <div>
            <h4
              class="ClockFormRow--heading"
            >
              When do you want to warn?
            </h4>
            <div
              class="ClockFormRow row "
            >
              <div
                class="column time"
              >
                <label
                  for="warningTime"
                >
                  Warning Time
                </label>
                <input
                  id="warningTime"
                  name="warningTime"
                  type="time"
                  value=""
                />
              </div>
              <div
                class="column"
              >
                <label
                  for="warningColor"
                >
                  Warning Color
                </label>
                <input
                  id="warningColor"
                  name="warningColor"
                  type="color"
                  value="#FFEC9E"
                />
              </div>
            </div>
          </div>
          <div>
            <h4
              class="ClockFormRow--heading"
            >
              When is time up?
            </h4>
            <div
              class="ClockFormRow row "
            >
              <div
                class="column time"
              >
                <label
                  for="endTime"
                >
                  End Time
                </label>
                <input
                  id="endTime"
                  name="endTime"
                  type="time"
                  value=""
                />
              </div>
              <div
                class="column"
              >
                <label
                  for="endColor"
                >
                  End Color
                </label>
                <input
                  id="endColor"
                  name="endColor"
                  type="color"
                  value="#FEADCD"
                />
              </div>
            </div>
          </div>
          <input
            class="ClockForm--button button hidden"
            type="submit"
            value="Create clock!"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
